File: javascript/nucleotide-count/nucleotide-count.js

function Strand(dnaStrand) {
  this.strand = dnaStrand;
}

Strand.prototype.count = function (nucleotide) {
  return this.strand.split(nucleotide).length - 1;
};

Strand.prototype.histogram = function () {
  return {
    A: this.count('A'),
    C: this.count('C'),
    G: this.count('G'),
    T: this.count('T')
  };
};

module.exports = function StrandFactory (dnaStrand) {
  var VALID_STRAND_NUCLEOTIDES = /[^ACGT]/,
    strand = dnaStrand || '';

  if (VALID_STRAND_NUCLEOTIDES.test(strand)) {
    throw new Error('DNA strand [' + strand + '] has no valid nucleotides');
  }

  return new Strand(strand);
};

File: elixir/nucleotide-count/nucleotide_count.exs
defmodule DNA do
  @nucleotides [ ?A, ?C, ?G, ?T ]

  @spec count([char], char) :: non_neg_integer
  def count(strand, nucleotide) when not nucleotide in @nucleotides, do: raise ArgumentError
  def count(strand, nucleotide) do
    Enum.count(strand, &(valid_nucleotide?(&1, nucleotide)))
  end

  defp valid_nucleotide?(s, _) when not s in @nucleotides, do: raise ArgumentError
  defp valid_nucleotide?(n, n), do: true
  defp valid_nucleotide?(_, _), do: false

  @spec histogram([char]) :: map
  def histogram(strand) do
    for n <- @nucleotides, into: Map.new do
      { n, count(strand, n) }
    end
  end
end
